state/executor.go

```diff
@@ -290,42 +319,120 @@ func (e *BlockExecutor) execute(ctx context.Context, state types.State, block *t
 		}
 	})
 
+	genAndGossipFraudProofIfNeeded := func(beginBlockRequest *abci.RequestBeginBlock, deliverTxRequests []*abci.RequestDeliverTx, endBlockRequest *abci.RequestEndBlock) (err error) {
+		if fraudProofsEnabled {
+			isr, err := e.getAppHash()
+			if err != nil {
+				return err
+			}
+			ISRs = append(ISRs, isr)
+			isFraud := e.checkFraudProofTrigger(isr, currentIsrs, currentIsrIndex)
+			if isFraud {
+				fraudProof, err := e.generateFraudProof(beginBlockRequest, deliverTxRequests, endBlockRequest)
+				if err != nil {
+					return err
+				}
+				// TODO: gossip fraudProof to P2P network
+				// fraudTx: current DeliverTx
+				_ = fraudProof
+			}
+			currentIsrIndex++
+		}
+		return nil
+	}
+
 	hash := block.Hash()
 	abciHeader, err := abciconv.ToABCIHeaderPB(&block.Header)
 	if err != nil {
 		return nil, err
 	}
 	abciHeader.ChainID = e.chainID
 	abciHeader.ValidatorsHash = state.Validators.Hash()
-	abciResponses.BeginBlock, err = e.proxyApp.BeginBlockSync(
-		abci.RequestBeginBlock{
-			Hash:   hash[:],
-			Header: abciHeader,
-			LastCommitInfo: abci.LastCommitInfo{
-				Round: 0,
-				Votes: nil,
-			},
-			ByzantineValidators: nil,
-		})
+	beginBlockRequest := abci.RequestBeginBlock{
+		Hash:   hash[:],
+		Header: abciHeader,
+		LastCommitInfo: abci.LastCommitInfo{
+			Round: 0,
+			Votes: nil,
+		},
+		ByzantineValidators: nil,
+	}
+	abciResponses.BeginBlock, err = e.proxyApp.BeginBlockSync(beginBlockRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	err = genAndGossipFraudProofIfNeeded(&beginBlockRequest, nil, nil)
 	if err != nil {
 		return nil, err
 	}
 
+	deliverTxRequests := make([]*abci.RequestDeliverTx, len(block.Data.Txs))
 	for _, tx := range block.Data.Txs {
-		res := e.proxyApp.DeliverTxAsync(abci.RequestDeliverTx{Tx: tx})
+		deliverTxRequest := abci.RequestDeliverTx{Tx: tx}
+		deliverTxRequests = append(deliverTxRequests, &deliverTxRequest)
+		res := e.proxyApp.DeliverTxAsync(deliverTxRequest)
 		if res.GetException() != nil {
 			return nil, errors.New(res.GetException().GetError())
 		}
+
+		err = genAndGossipFraudProofIfNeeded(&beginBlockRequest, deliverTxRequests, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+	endBlockRequest := abci.RequestEndBlock{Height: int64(block.Header.Height)}
+	abciResponses.EndBlock, err = e.proxyApp.EndBlockSync(endBlockRequest)
+	if err != nil {
+		return nil, err
 	}
 
-	abciResponses.EndBlock, err = e.proxyApp.EndBlockSync(abci.RequestEndBlock{Height: int64(block.Header.Height)})
+	err = genAndGossipFraudProofIfNeeded(&beginBlockRequest, deliverTxRequests, &endBlockRequest)
 	if err != nil {
 		return nil, err
 	}
 
+	if fraudProofsEnabled && block.Data.IntermediateStateRoots.RawRootsList == nil {
+		// Block producer: Initial ISRs generated here
+		block.Data.IntermediateStateRoots.RawRootsList = ISRs
+	}
+
 	return abciResponses, nil
 }
 
+func (e *BlockExecutor) checkFraudProofTrigger(generatedIsr []byte, currentIsrs [][]byte, index int) bool {
```

Comment:
Since this function returns a boolean, I think it is more idiomatic to be `isFraudProofTrigger` the `check` prefix makes me think that it should be returning an error.